jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageReference.java

```diff
@@ -31,53 +32,73 @@
  */
 public class ImageReference {
 
+  /** Regular expression patterns for the image reference. */
+  private static class Regex extends RegexBuilder {
+
+    /**
+     * Matches all sequences of alphanumeric characters possibly separated by any number of dashes
+     * in the middle.
+     */
+    private static final String REGISTRY_PART =
+        group(alphanum(), group(any(alphanum('-')), alphanum()));
+
+    /**
+     * Matches sequences of {@link #REGISTRY_PART} separated by a dot, with an optional {@code
+     * :port} at the end.
+     */
+    private static final String REGISTRY =
+        sequence(
+            REGISTRY_PART,
+            any(group(literal('.'), REGISTRY_PART)),
+            optional(group(literal(':'), repeated(digit()))));
+
+    /**
+     * Matches all sequences of lowercase alphanumeric characters separated by a separator.
+     *
+     * <p>A separator is either an underscore, a dot, two underscores, or any number of dashes.
+     */
+    private static final String REPOSITORY_PART =
+        sequence(
```

Comment:
I found this one to be pretty confusing. Could this expression be simplified? Or, I wonder if some of these could be more clear by using a Regex-DSL hybrid rather than just one or the other.